### PR TITLE
Added support for time attributes

### PR DESCRIPTION
**Pull request by andrzejdackiewicz**

### CODE REVIEW

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java [hunk 1]
```diff
@@ -56,29 +57,22 @@ public class ElasticCatalog extends AbstractJdbcCatalog {
...
         this(catalogName, defaultDatabase, username, password, baseUrl, new HashMap<>());
     }
 
-    public ElasticCatalog(String catalogName,
-                          String defaultDatabase,
-                          String username,
-                          String password,
-                          String baseUrl,
-                          Map<String, String> properties) {
+    public ElasticCatalog(String catalogName, String defaultDatabase, String username, String password, String baseUrl, Map<String, String> properties) {
```
Comment (grzegorz8) [inline flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java]: Please revert it. The line is too long and imho it's better to add line breaks

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java [hunk 2]
```diff
@@ -97,35 +91,25 @@ private String[] extractCatalogDefaultScanProperties(Map<String, String> propert
...
         // Splitting patterns and removing duplicates
 
-        return Arrays.stream(
-                        properties.getOrDefault("properties.index.patterns", "").split(",")
-                )
-                .map(String::trim)
-                .filter(e -> !e.isEmpty())
-                .distinct()
-                .collect(Collectors.toList());
+        return Arrays.stream(properties.getOrDefault("properties.index.patterns", "").split(",")).map(String::trim).filter(e -> !e.isEmpty()).distinct().collect(Collectors.toList());
```
Comment (grzegorz8) [inline flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java]: Please revert this change too.

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java [hunk 3]
```diff
@@ -97,35 +91,25 @@ private String[] extractCatalogDefaultScanProperties(Map<String, String> propert
...
             }
-            String tableName = key.replace("properties.scan.", "")
-                    .replace(".partition.column.name", "")
-                    .replace(".partition.number", "");
+            String tableName = key.replace("properties.scan.", "").replace(".partition.column.name", "").replace(".partition.number", "");
             boolean scanPropertiesForTableFound = scanPartitionProperties.containsKey(tableName);
-            ScanPartitionProperties partitionProperties = scanPropertiesForTableFound
-                    ? scanPartitionProperties.get(tableName)
-                    : new ScanPartitionProperties();
+            ScanPartitionProperties partitionProperties = scanPropertiesForTableFound ? scanPartitionProperties.get(tableName) : new ScanPartitionProperties();
```
Comment (grzegorz8) [inline flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java]: Please try no to exceed 120 line width. Please apply it everywhere.

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java [hunk 4]
```diff
@@ -139,12 +123,47 @@ private Map<String, ScanPartitionProperties> extractScanTablePartitionProperties
...
+                    replace(".watermark.column", "").
+                    replace(".watermark.delay", "").
+                    replace(".proctime.column", "");
+            boolean timeAttributePropertiesForTableFound = timeAttributeProperties.containsKey(tableName);
+            TimeAttributeProperties timeAttributePropertiesForTable = timeAttributePropertiesForTableFound ? timeAttributeProperties.get(tableName) : new TimeAttributeProperties();
+
+            if (entry.getKey().endsWith(".watermark.column")) {
+                if (timeAttributePropertiesForTable.getWatermarkColumn() == null) {
+                    timeAttributePropertiesForTable.setWatermarkColumn(entry.getValue());
+                }
```
Comment (grzegorz8) [inline flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java:173]: what if != null? x3

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java [hunk 5]
```diff
@@ -253,7 +270,26 @@ private void deducePartitionNumber(ScanPartitionProperties properties, Connectio
...
+        String watermarkDelay = timeAttributeProperties.getWatermarkDelay();
+        String watermarkColumn = timeAttributeProperties.getWatermarkColumn();
+
+        if (proctimeColumn != null && (
+                watermarkDelay != null ||
+                        watermarkColumn != null)) {
+            throw new IllegalArgumentException("Either proctime or watermark properties should be specified for a table " + tableName + ".");
+        }
+        if (proctimeColumn == null && (watermarkDelay == null || watermarkColumn == null)) {
+            throw new IllegalArgumentException("You should specify both watermarkDelay and watermakColumn properties when using watermark for a table " + tableName + ".");
```
Comment (grzegorz8) [inline flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java]: typo: watermak

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java [hunk 6]
```diff
@@ -453,10 +489,49 @@ public void setScanPartitionUpperBound(Long upperBound) {
...
+            return "partitionColumnName=" + partitionColumnName + ", partitionNumber=" + partitionNumber + ", scanPartitionLowerBound=" + scanPartitionLowerBound + ", scanPartitionUpperBound=" + scanPartitionUpperBound;
+        }
+    }
+
+    static class TimeAttributeProperties {
+        private String watermarkColumn;
+        private String watermarkDelay;
+        private String proctimeColumn;
+
+        public TimeAttributeProperties() {
```
Comment (grzegorz8) [inline flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java]: You dont need to define default constructor which sets all fields to null. If no constructor is defined, the default one is created.

File: flink-elastic-catalog/src/test/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalogTest.java [hunk 7]
```diff
@@ -577,21 +577,137 @@ public void testGetTablePartitionBySpecialCharacterColumn() throws TableNotExist
...
+        // when
+        ElasticCatalog catalog = new ElasticCatalog("test-catalog", "test-database", USERNAME, PASSWORD, url, properties);
+        CatalogBaseTable table = catalog.getTable(new ObjectPath("docker-cluster", "test_multiple_records_table"));
+
+        // then
+        Schema schema = table.getUnresolvedSchema();
+        assertNotNull(table);
+        assertNotNull(schema);
+
+        Schema expectedSchema = Schema.newBuilder().fromFields(
```
Comment (grzegorz8) [inline flink-elastic-catalog/src/test/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalogTest.java]: Please check only if proctime column is defined. I think there is no need to check the entire schema

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java (hunk 5 above)
Comment (andrzejdackiewicz) [reply flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java]: Fixed!

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java (hunk 6 above)
Comment (andrzejdackiewicz) [reply flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java]: Yes, in this case we do not need the constructor :) Thanks!

File: flink-elastic-catalog/src/test/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalogTest.java (hunk 7 above)
Comment (andrzejdackiewicz) [reply flink-elastic-catalog/src/test/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalogTest.java]: Fixed :)

File: flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java (hunk 4 above)
Comment (andrzejdackiewicz) [reply flink-elastic-catalog/src/main/java/com/getindata/flink/connector/jdbc/catalog/ElasticCatalog.java:173]: If all 3 timeAttribute properties are specified then `checkTimeAttributeProperties` method should raise an error saying that `Either proctime or watermark properties should be specified for a table` there is a unit test `testFailGetTableTimeAttributesProctimeAndWatermarkProvided` covering this case.

**Review by grzegorz8 — APPROVED** [review]
